// @nestjs/typeorm/dist/common/typeorm.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@nestjs/common");
const operators_1 = require("rxjs/operators");
const typeorm_1 = require("typeorm");
const uuid = require("uuid/v4");
const logger = new common_1.Logger('TypeOrmModule');
function getRepositoryToken(entity) {
    if (entity.prototype instanceof typeorm_1.Repository ||
        entity.prototype instanceof typeorm_1.AbstractRepository) {
        return getCustomRepositoryToken(entity);
    }
    return `${entity.name}Repository`;
}
exports.getRepositoryToken = getRepositoryToken;
function getCustomRepositoryToken(repository) {
    return repository.name;
}
exports.getCustomRepositoryToken = getCustomRepositoryToken;
function getConnectionToken(connection = 'default') {
    return 'default' === connection
        ? typeorm_1.Connection
        : 'string' === typeof connection
            ? `${connection}Connection`
            : 'default' === connection.name || !connection.name
                ? typeorm_1.Connection
                : `${connection.name}Connection`;
}
exports.getConnectionToken = getConnectionToken;
function getEntityManagerToken(connection = 'default') {
    return 'default' === connection
        ? typeorm_1.EntityManager
        : 'string' === typeof connection
            ? `${connection}EntityManager`
            : 'default' === connection.name || !connection.name
                ? typeorm_1.EntityManager
                : `${connection.name}EntityManager`;
}
exports.getEntityManagerToken = getEntityManagerToken;
function handleRetry(retryAttempts = 9, retryDelay = 3000) {
    return (source) => source.pipe(operators_1.retryWhen(e => e.pipe(operators_1.scan((errorCount, error) => {
        logger.error(`Unable to connect to the database. Retrying (${errorCount +
            1})...`, error.stack);
        if (errorCount + 1 >= retryAttempts) {
            throw error;
        }
        return errorCount + 1;
    }, 0), operators_1.delay(retryDelay))));
}
exports.handleRetry = handleRetry;
function getConnectionName(options) {
    return options && options.name ? options.name : 'default';
}
exports.getConnectionName = getConnectionName;
exports.generateString = () => uuid();
